In [1]:
#Importing libraries
import pandas as pd
import numpy as np
import os

In [2]:
# Defining path
path = r'/Users/CK/Downloads/CF/Instacart Basket Analysis'



In [3]:
# Importing customer dataset
customer_data = pd.read_csv(os.path.join(path, '02 Data' , 'Original Data', 'customers.csv'))

In [4]:
# Checking a preview of the dataset
customer_data.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

# Q4 Wrangle the data so that it follows consistent logic; for example, rename columns with illogical names and drop columns that donot add anything to your analysis

In [5]:
# Renaming columns
customer_data.rename(columns={'Surnam':'surname', 'fam_status':'family_status','STATE':'State','n_dependants':'dependants','First Name':'first_name'}, inplace=True)

In [6]:
# Checking a preview of the dataset
customer_data.head()

user_id first_name   surname  Gender       State  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   dependants family_status  income  
0           3       married  165665  
1           0        single   59285  
2           2       married   99568  
3           0        single   42049  
4           1       married   40374

# Complete the fundamental data quality and consistency checks you’ve learned throughout this Achievement; for example, check for and address missing values and duplicates, and convert any mixed-type data.

In [7]:
# Check duplicate data
customer_data_dups = customer_data[customer_data.duplicated()]

In [8]:
customer_data_dups

Empty DataFrame
Columns: [user_id, first_name, surname, Gender, State, Age, date_joined, dependants, family_status, income]
Index: []

##There are no duplicate data

In [9]:
# check missing values
customer_data.isnull().sum()

user_id              0
first_name       11259
surname              0
Gender               0
State                0
Age                  0
date_joined          0
dependants           0
family_status        0
income               0
dtype: int64

There are 11259 missing values in the first_name column. Nothing would be done here and all the records will be kept as it is.


In [10]:
# check datatypes
customer_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206209 entries, 0 to 206208
Data columns (total 10 columns):
 #   Column         Non-Null Count   Dtype 
---  ------         --------------   ----- 
 0   user_id        206209 non-null  int64 
 1   first_name     194950 non-null  object
 2   surname        206209 non-null  object
 3   Gender         206209 non-null  object
 4   State          206209 non-null  object
 5   Age            206209 non-null  int64 
 6   date_joined    206209 non-null  object
 7   dependants     206209 non-null  int64 
 8   family_status  206209 non-null  object
 9   income         206209 non-null  int64 
dtypes: int64(4), object(6)
memory usage: 15.7+ MB


In [11]:
# change datejoined as date datatype
customer_data['date_joined']=pd.to_datetime(customer_data['date_joined'])

In [12]:
customer_data.dtypes

user_id                   int64
first_name               object
surname                  object
Gender                   object
State                    object
Age                       int64
date_joined      datetime64[ns]
dependants                int64
family_status            object
income                    int64
dtype: object

## Q6) Combine your customer data with the rest of your prepared Instacart data. (Hint: Make sure the key columns are the same data type!)


In [13]:
# Importing order_products_grouped data set
df_orderdata = pd.read_pickle(os.path.join(path, '02 Data' , 'Prepared Data', 'ords_prods_merge_upd_grouped.pkl'))

In [14]:
df_orderdata.dtypes

order_id                     int64
user_id                      int64
order_number                 int64
orders_day_of_week           int64
order_hour_of_day            int64
days_since_prior_order     float64
product_id                   int64
add_to_cart_order            int64
reordered                    int64
product_name                object
aisle_id                     int64
department_id                int64
prices                     float64
_merge                    category
price_range_loc             object
busiest_days                object
busiest_day                 object
busiest_period_of_day       object
max_order                    int64
loyalty_flag                object
mean_product_price         float64
spender_type                object
order_frequency            float64
frequency_flag              object
dtype: object

In [15]:
df_orderdata=df_orderdata.drop('_merge', axis= 1)

In [16]:
df_orderdata.dtypes

order_id                    int64
user_id                     int64
order_number                int64
orders_day_of_week          int64
order_hour_of_day           int64
days_since_prior_order    float64
product_id                  int64
add_to_cart_order           int64
reordered                   int64
product_name               object
aisle_id                    int64
department_id               int64
prices                    float64
price_range_loc            object
busiest_days               object
busiest_day                object
busiest_period_of_day      object
max_order                   int64
loyalty_flag               object
mean_product_price        float64
spender_type               object
order_frequency           float64
frequency_flag             object
dtype: object

In [17]:
# The key column user_id has same datatype in both the dataframes which is good. Next, we combine dataframes on user_id
df_merged=df_orderdata.merge(customer_data,on='user_id',indicator=True)

## Export data to pkl



In [22]:
df_merged.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'customer_and_order.pkl'))